torch_tvm/compiler.cpp

```diff
@@ -368,27 +449,50 @@ void TVMCompiler::run(Stack& stack) {
   std::unordered_map<Value*, IValue> value_to_ivalue;
   int num_inputs = subgraph_->inputs().size();
   at::ArrayRef<IValue> inputs = last(stack, num_inputs);
+  CompleteArgumentSpec spec{false, ArrayRef<IValue>(inputs)};
+
+  if (bad_specs_.count(spec)) {
+    fallback_state_->run(stack);
+    return;
+  }
 
+#ifdef TVM_TRAINING
+  bool is_all_cuda = true;
   for (auto i = 0; i < inputs.size(); ++i) {
     auto value_input = subgraph_->inputs()[i];
     value_to_ivalue[value_input] = inputs[i];
+    if (inputs[i].isTensor()) {
+      if (!inputs[i].toTensor().device().is_cuda()) {
+        is_all_cuda = false;
+      }
+    }
   }
 
-  CompleteArgumentSpec spec{false, ArrayRef<IValue>(inputs)};
+  if (!is_all_cuda) {
+    bad_specs_.insert(spec);
+    fallback_state_->run(stack);
+    return;
+  }
+#endif
 
   if (cache_.find(spec) == cache_.end()) {
     for (auto& kv : value_to_ivalue) {
       if (kv.second.isTensor()) {
         kv.first->inferTypeFrom(kv.second.toTensor());
       } else if (kv.second.isInt()) {
         kv.first->setType(IntType::get());
+      } else if (kv.second.isDouble()) {
+        kv.first->setType(FloatType::get());
       } else {
-        AT_CHECK(
-            0,
-            "Cannot handle this type yet ",
-            kv.second,
-            "\nGraph:\n",
-            *subgraph_);
+        std::cout << "Cannot handle this type yet "
```

Comment:
let's use a logging utility (there are a couple in PT) rather than std::cout

Reply:
I'll check